### ISSUE TITLE
Have tilt-prepare fetch the latest version of a Helm chart in `hack/observability`

### ISSUE THREAD

**Issue by Jont828** [kubernetes-sigs/cluster-api#7052]
**User Story**

Currently, the folder `hack/observability/*/charts` contains cached Helm charts. I'm running into an issue where if tilt-prepare fetches a Helm chart and later on a new version is released, tilt-prepare will still use the cached chart. In order to fetch the latest chart, I need to delete the charts folder. I would like tilt-prepare to be able to update the cache if there is a new version of the chart available.

/kind feature

**Comment by Jont828** [comment]
@fabriziopandini @sbueringer

**Comment by sbueringer** [comment]
Sounds reasonable to me. Not sure how to do it though, without pulling it every time.

Probably we have to look into how `helmCharts` are implemented in kustomize?

**Comment by Jont828** [comment]
@sbueringer It looks like [here](https://github.com/kubernetes-sigs/cluster-api/blob/fa70f61a77828cea2d5d059ffd0dfcf6bcbc9f2d/hack/tools/tilt-prepare/main.go#L496) is where we're getting the Helm chart using kustomize. Would there be a flag like `--no-cache` to force it to repull the chart? Or would we want to just delete the charts in the Makefile before triggering the tilt-prepare code?

**Comment by sbueringer** [comment]
We are just running kustomize and kustomize then pulls the Chart based on e.g. this configuration: https://github.com/kubernetes-sigs/cluster-api/blob/34676f0670506e70a61b0f2166768ca3281be509/hack/observability/prometheus/kustomization.yaml

I'm not sure how a `no-cache` flag would work. I would want to avoid that we have to pull the Charts on every `tilt up`. I don't know if the kustomize Helm configuration already has an option to always pull the latest Chart if the local one is outdated.

**Comment by Jont828** [comment]
Yeah I think it's a little confusing that the `version` flag says it defaults to the latest version if not set but won't update if a new version comes out. I suppose I can wait for a response to the issue I opened for kustomize.

**Comment by fabriziopandini** [comment]
given that this is a dev feature I'm ok to have a meke target to cleanup the chart folder on demand + some documentation somewhere

**Comment by ykakarap** [comment]
Kustomize does not check for version changes even if the verison is specific or modified later as long as there is a local copy of a chart (any version). 

From Kustomize documentation:

>If a chart exists, kustomize will not overwrite it (so to suppress a pull, simply assure the chart is already in your kustomization root). kustomize won't check dates or version numbers or do anything that smells like cache management.


If we pin the chart version in the our kustomization.yaml files then in tilt-preare we can be a little smart by checking if the version in `Chart.yaml` matches the version specified in the corresponding kustomization.yaml. Example: check for version in `hack/observability/prometheus/charts/prometheus/Chart.value` with version in `hack/observability/prometheus/kustomization.yaml`.
This way we wont have to pull the chart everytime.

**Comment by sbueringer** [comment]
To be honest I'm not sure if it's worth the effort. We're pulling in external charts for relatively stable components like Prometheus/Grafana/Loki. 

If we start pinning, we also have to regularly bump those versions (and probably test them as well, because otherwise pinning doesn't really have a big advantage).

As of today - as far as I'm aware - we had 0 issues with users of our dev env complaining about broken observability Charts / mismatched versions, so I wonder how much effort we should invest in this (apart from the cleanup target).

**Comment by killianmuldoon** [comment]
>As of today - as far as I'm aware - we had 0 issues with users of our dev env complaining about broken observability charts / mismatched versions, so I wonder how much effort we should invest in this (apart from the cleanup target).

@ykakarap and I have run into an issue with Loki/promtail being improperly configured which seemed down to the version of the chart being pulled. Promtail was trying to contact the wrong service to push to Loki.

**Comment by sbueringer** [comment]
If we start pinning it would be really good to have a rough idea what kind of tests we want to do when bumping Charts. (pinning and manual bumping without any tests wouldn't really improve the situation just produce more toil)

**Comment by fabriziopandini** [comment]
/triage accepted

This is the same problem we have for all the tools we build locally, and for those we are ok with a make target to cleanup them only when necessary. For consistency, I will do the same for helm charts (and avoid the burden of pinning versions)

**Comment by fabriziopandini** [comment]
/close
For now we are not pinning releases to avoid maintenance burden.
We will reconsider only if we detect that some of the charts we are using will introduce breaking changes repeatedly
